api/service/model_endpoint_service.go

```diff
@@ -124,6 +124,16 @@ func (s *modelEndpointsService) DeployEndpoint(ctx context.Context, model *model
 		return nil, fmt.Errorf("unable to find istio client for environment: %s", endpoint.EnvironmentName)
 	}
 
+	// if there's error during VirtualService creation, we will rollback the changes, to make sure
+	// there's no hanging state of vs and to align with the database state
+	defer func() {
+		if err != nil {
+			if err := istioClient.DeleteVirtualService(context.WithoutCancel(ctx), model.Project.Name, model.Name); err != nil {
+				log.Errorf("failed to rollback create VirtualService: %v", err)
+			}
+		}
+	}()
```

Comment:
Ahhh so if I understand correctly, we're deleting this virtual service if any of the steps following (and including) the virtual service creation fails? i.e. if any of the following functions return an error,
- `istioClient.CreateVirtualService`
- `s.modelEndpointStorage.Save`
- `s.observabilityEventProducer.ModelEndpointChangeEvent`

we delete the virtual service created? Wow this might actually be the cause of the bug 🤔 Like improper virtual service clean up because the whole set of operations aren't atomic (we also aren't reversing the other steps that have been executed like the DB update when subsequent functions fail). Thanks for digging through the code and fixing the workflow!